src/chainstate/stacks/index/file.rs

```diff
@@ -0,0 +1,480 @@
+        if path == ":memory:" {
+            Ok(TrieFile::new_ram(readonly))
+        } else {
+            let blob_path = format!("{}.blobs", path);
+            TrieFile::new_disk(&blob_path, readonly)
+        }
+    }
+
+    /// Write a trie blob to external storage, and add the offset and length to the trie DB.
+    /// Return the trie ID
+    pub fn store_trie_blob<T: MarfTrieId>(
+        &mut self,
+        db: &Connection,
+        bhh: &T,
+        buffer: &[u8],
+        block_id: Option<u32>,
+    ) -> Result<u32, Error> {
+        let offset = self.append_trie_blob(db, buffer)?;
+        test_debug!("Stored trie blob {} to offset {}", bhh, offset);
+        trie_sql::write_external_trie_blob(db, bhh, offset, buffer.len() as u64, block_id)
+    }
+
+    /// Read a trie blob in its entirety from the DB
+    fn read_trie_blob_from_db(db: &Connection, block_id: u32) -> Result<Vec<u8>, Error> {
+        let trie_blob = {
+            let mut fd = trie_sql::open_trie_blob_readonly(db, block_id)?;
+            let mut trie_blob = vec![];
+            fd.read_to_end(&mut trie_blob)?;
+            trie_blob
+        };
+        Ok(trie_blob)
+    }
+
+    /// Read a trie blob in its entirety from the blobs file
+    #[cfg(test)]
+    pub fn read_trie_blob(&mut self, db: &Connection, block_id: u32) -> Result<Vec<u8>, Error> {
+        let (offset, length) = trie_sql::get_external_trie_offset_length(db, block_id)?;
+        self.seek(SeekFrom::Start(offset))?;
+
+        let mut buf = vec![0u8; length as usize];
+        self.read_exact(&mut buf)?;
+        Ok(buf)
+    }
+
+    /// Copy the trie blobs out of a sqlite3 DB into their own file
+    pub fn export_trie_blobs<T: MarfTrieId>(&mut self, db: &Connection) -> Result<(), Error> {
+        let max_block = trie_sql::count_blocks(db)?;
+        info!(
+            "Migrate {} blocks to external blob storage at {}",
+            max_block,
+            &self.get_path()
+        );
+        for block_id in 0..(max_block + 1) {
+            match trie_sql::is_unconfirmed_block(db, block_id) {
+                Ok(true) => {
+                    test_debug!("Skip block_id {} since it's unconfirmed", block_id);
+                    continue;
+                }
+                Err(Error::NotFoundError) => {
+                    test_debug!("Skip block_id {} since it's not a block", block_id);
+                    continue;
+                }
+                Ok(false) => {
+                    // get the blob
+                    let trie_blob = TrieFile::read_trie_blob_from_db(db, block_id)?;
+
+                    // get the block ID
+                    let bhh: T = trie_sql::get_block_hash(db, block_id)?;
+
+                    // append the blob, replacing the current trie blob
+                    info!(
+                        "Migrate block {} ({} of {}) to external blob storage",
+                        &bhh, block_id, max_block
+                    );
```

Comment:
This should maybe only occur on `block_id % 1000 == 0` or something like that -- otherwise the migration will be 55k log lines.

Reply:
Yup, will fix.